File: code/STEP4_5/convert1.HadR2_mod4.f
c   To read the  oi SST version 2
c    and the masks.
c
c
      parameter (im=360,jm=180,ioff=im/2)
      parameter (iyrbeg=1880,iyreo=3000, nymax=15)
      parameter (monx=12*(iyreo-iyrbeg+1))
c
c   i is longitude index (values 1 to im)
c   j is the latitude index (values 1 to jm)
c
      real*4 sst(im,jm,12*nymax)
c
c   sst is the input SST real*2 array
c    (values are degrees C)
c
      integer info(8)
c
c   ls is the land/sea array not needed - built into climatology
c
      real to(monx)
      real clim(im,jm,12)
c
c   rsst is the reconstructed SST mean in degrees C
c   Read the EOF reconstruction area mask (unit 4)
c
      character*80 filei/
     *  'input/oiv2mon.yyyymm'/,line,title
      ny=index(filei,'yyyy')

      open(1,file='input/SBBX.HadR2',form='unformatted')
      read(1) info,title
      bad=info(7)
      mnow=info(1)
      info(1)=1  ! output file NOT trimmed/reorganized
      if(info(6).ne.iyrbeg) then
        write(*,*) 'iyrbeg-new',iyrbeg,' iyrbeg-old',info(6)
        stop 'iyrbeg inconsistent'
      end if
      write(*,*) 'SBBX.HadR2  opened'
      open(24,file='input/SBBX_LtSN.LnWE.dat',form='unformatted')

      open(14,file='input/oisstv2_mod4.clim',
     *        form='unformatted')
      read(14) line,clim
      close (14)
c
c   Read in the SST data for recent years (unit 11)
c
      call getarg(1,line)
      read(line,*) iyr1
      if(iyr1.lt.iyrbeg) then
        write(*,*) 'iyr1=',iyr1,' cannot be less than',iyrbeg
        stop 'arg 1 off'
      end if
      if(iyr1.gt.iyreo) then
        write(*,*) 'iyr1=',iyr1,' cannot be less than',iyreo
        stop 'arg 1 off'
      end if
      call getarg(2,line)
      read(line,*) mo1
      if(mo1.lt.1.or.mo1.gt.12) then
        write(*,*) 'mo1=',mo1,' must be between 1 and 12'
        stop 'arg 2 off'
      end if
      call getarg(3,line)
      read(line,*) mo2
      if(mo2.lt.mo1.or.mo2.gt.12) then
        write(*,*) 'mo2=',mo2,' must be between',mo1,' and 12'
        stop 'arg 3 off'
      end if
      info(4)=12*(iyr1-iyrbeg+1) ! wipe out data after new data
      info(5)=info(4)+4     ! old non-trimmed format

      do 100 iyr=iyr1,iyr1
      do 90 mon=mo1,mo2
      write(filei(ny:ny+5),'(I4.4,I2.2)') iyr,mon
      write(*,*) 'trying to read ',filei(1:70)
      open(11,form='unformatted',file=filei,err=110)
      read(11) iyr0,imo
      iyre=iyr
      moe=mon
      if(iyr.ne.iyr0) write(*,*) 'years not ok',iyr,iyr0
      if(mon.ne.imo) write(*,*) 'months not ok',mon,imo
c
c   iyr is year (value 1950 to ....)
c   imo is month (value 1 to 12)
c
      read(11) ((sst(i,j,12*(iyr-iyr1)+mon),i=1,im),j=1,jm)
   90 close (11)
  100 continue
  110 open(2,file='work/SBBX.HadR2.upd',form='unformatted')
      title(41:80)=' Had: 1880-11/1981, oi2: 12/1981-mm/yyyy'
      write(title(74:80),'(I2,''/'',I4)') moe,iyre
      write(2) info,title
c****
c**** Interpolate to Sergei's subbox grid
c****
      moff=(iyr1-iyrbeg)*12
      do 500 n=1,8000
      do 200 m=1,monx
  200 to(m)=bad
      call sread(1,to,mnow,lts,ltn,lnw,lne,mnext)
      mnow=mnext
      js=(18001+(lts+9000)*jm)/18000
      jn=(17999+(ltn+9000)*jm)/18000
      iw=(36001+(lnw+18000)*im)/36000+ioff
      ie=(35999+(lne+18000)*im)/36000+ioff
      if(ie.gt.im) then
         iw=iw-im
         ie=ie-im
      end if
      if(iw.gt.im) stop 'iw>im'
      if(ie.gt.im) stop 'ie>im'
      if(iw.lt.1) stop 'iw<1'
      if(ie.lt.1) stop 'ie<1'


      do 220 m=mo1,mo2
      month=mod(m-1,12)+1
      kt=0
      ts=0.
      do 210 j=js,jn
      do 210 i=iw,ie
      if(sst(i,j,m).le.-1.77.or.clim(i,j,month).eq.bad) go to 210
      kt=kt+1
      ts=ts+(sst(i,j,m)-clim(i,j,month))
c     write(*,*) 'new data:',m,' at',i,j,n
  210 continue
        to(m+moff)=bad
        if(kt.gt.0) to(m+moff)=ts/kt
  220 continue
      do 230 m=(iyre-iyrbeg)*12+moe+1,info(4)
c     write(*,*) 'filling with bad:',m
  230 to(m)=bad

      write(2) (to(i),i=1,info(4)),lts,ltn,lnw,lne
  500 continue
      stop
      end

      subroutine sread(in,a,na,lts,ltn,lnw,lne,mnext)
      real a(na)
      integer info(7)
      read(in) mnext,lts,ltn,lnw,lne,x,x,x,a
      return
      end

File: code/convert1.HadR2_mod4.py
#!/usr/bin/env python
"""Python version of convert1.HadR2_mod4.py.

"""
__docformat__ = "restructuredtext"


import sys
import string
import struct
import array
import copy

import fort
import ccc_binary
import script_support


# Constants (from original fortran parameters)
IM = 360
JM = 180
IOFF = IM // 2
IYRBEG = 1880
IYREO = 3000
NYMAX = 15
MONX = 12 * (IYREO - IYRBEG + 1)


def make_3d_array(a, b, c):
    """Create an array with three dimensions.

    Actually a list-of-lists-of-lists, but the result can be treated as an
    array.

    """
    x = [0.0] * c
    arr = []
    for i in range(a):
        arr.append([list(x) for j in range(b)])

    return arr


def sread(f, na):
    data = f.readline()
    mnext, lts, ltn, lnw, lne, x, x, x = struct.unpack(">lllllfff", data[:32])
    a = struct.unpack(">%df" % na, data[32:])
    return mnext, lts, ltn, lnw, lne, a


def load_oisstv2_mod4_clim():
    f14 = fort.open("input/oisstv2_mod4.clim")
    try:
        f14.bos = ">"
        data = f14.readline()
    finally:
        f14.close()
    clim_title, data = data[:80], data[80:]

    clim = make_3d_array(IM, JM, 12)
    ii = jj = kk = 0
    for p in range(0, len(data), 4):
        v, = struct.unpack(">f", data[p:p+4])
        int_v, = struct.unpack("<L", data[p:p+4])
        clim[ii][jj][kk] = v
        ii += 1
        if ii >= IM:
            ii = 0
            jj += 1
            if jj >= JM:
                jj = 0
                kk += 1

    return clim


def load_sst_data(iyr1, mo1, mo2):
    # Read in the SST data for recent years (unit 11)
    sst = make_3d_array(IM, JM, 12 * NYMAX)

    for iyr in range(iyr1, iyr1 + 1):
        for mon in range(mo1, mo2 + 1):
            path = "input/oiv2mon.%04d%02d" % (iyr, mon)
            sys.stdout.write(" trying to read %-70s\n" % path)
            f11 = fort.open(path)
            f11.bos = ">"
            data = f11.readline()
            iyr0, imo = struct.unpack(">LL", data[:8])
            iyre = iyr
            moe = mon
            if iyr != iyr0:
                sys.stderr.write("years not ok %s %s\n" % (iyr, iyr0))
            if mon != imo:
                sys.stderr.write("months not ok %s %s\n" % (mon, imo))

            data = f11.readline()
            f11.close()
            kk = 12 * (iyr - iyr1) + mon - 1
            ii = jj = 0
            for p in range(0, len(data), 4):
                v, = struct.unpack(">f", data[p:p+4])
                int_v, = struct.unpack("<L", data[p:p+4])
                sst[ii][jj][kk] = v
                ii += 1
                if ii >= IM:
                    ii = 0
                    jj += 1

    return sst, moe, iyre


def open_curr_SBBX():
    f1 = fort.open("input/SBBX.HadR2")
    f1.bos = ">"
    inH = ccc_binary.SBBX_Header(f1.readline(), bos=f1.bos)
    print inH.title

    return f1, inH


def main(args):
    f1, inH = open_curr_SBBX()

    bad = inH.bad
    xbad = float(bad)
    mnow = inH.m1

    if inH.yrbeg != IYRBEG:
        sys.stderr.write("Error: iyrbeg-new %s iyrbeg-old %s\n" % (
            IYRBEG, inH.yrbeg))
        sys.exit("Error: iyrbeg inconsistent")
    sys.stdout.write(" SBBX.HadR2  opened\n")

    # Paul: I do not know what this file is for. The fortran code opens the
    # file but does not seem to use it.
    # f24 = fort.open("input/SBBX_LtSN.LnWE.dat")
    # f24.bos = ">"

    iyr1, mo1, mo2 = [int(v) for v in args]
    clim = load_oisstv2_mod4_clim()
    sst, moe, iyre = load_sst_data(iyr1, mo1, mo2)

    outH = ccc_binary.SBBX_Header(bos="<")
    for attr in "m1 kq mavg nm recsize yrbeg bad trace".split():
        setattr(outH, attr, getattr(inH, attr))
    outH.m1 = 1                        # output file NOT trimmed/reorganized
    outH.nm = 12 * (iyr1 - IYRBEG + 1) # wipe out data after new data
    outH.recsize = outH.nm + 4         # old non-trimmed format
    outH.title = inH.title[:40] + " Had: 1880-11/1981, oi2: 12/1981-%2d/%04d" % (
            moe, iyre)

    f2 = fort.open("work/SBBX.HadR2.upd_py", "wb")
    f2.writeline(outH.binary)

    #
    # Interpolate to Sergei's subbox grid
    #
    to_ = [0.0] * MONX
    moff = (iyr1 - IYRBEG) * 12
    for n in range(8000):
        for m in range(MONX):
            to_[m] = bad
        mnext, lts, ltn, lnw, lne, a = sread(f1, mnow)
        to_[:mnow] = a
        mnow = mnext
        js = (18001 + (lts + 9000) * JM)/18000
        jn = (17999 + (ltn + 9000) * JM)/18000
        iw = (36001 + (lnw + 18000) * IM)/36000 + IOFF
        ie = (35999 + (lne + 18000) * IM)/36000 + IOFF
        if ie > IM:
            iw = iw - IM
            ie = ie - IM
        if iw > IM:
            sys.exit('Fatal: iw>im')
        if ie > IM:
            sys.exit('Fatal: ie>im')
        if iw < 1:
            sys.exit('Fatal: iw<1')
        if ie < 1:
            sys.exit('Fatal: ie<1')

        for m in range(mo1, mo2 + 1):
            month = (m - 1) % 12
            kt = 0
            ts = 0.0
            for j in range(js -1, jn):
                for i in range(iw - 1, ie):
                    if sst[i][j][m-1] < -1.77 or clim[i][j][month] == xbad:
                        continue
                    kt += 1
                    ts += sst[i][j][m-1] - clim[i][j][month]
            
            to_[m + moff - 1] = bad
            if kt > 0:
                to_[m + moff - 1] = ts / kt

        b = outH.nm
        data = to_[:b]
        data.extend([lts, ltn, lnw, lne])
        s = struct.pack("%dfllll" % (b,), *data)
        f2.writeline(s)


if __name__ == "__main__":
    import optparse
    usage = "usage: %prog [options] year month1 month2"
    parser = script_support.makeParser(usage)
    options, args = script_support.parseArgs(parser, __doc__, (3, 3))
    main(args)
